### PR TITLE
opencv: misc CUDA-related updates and fixes; add enableLto

### PR DESCRIPTION
**Pull request by ConnorBaker**
###### Description of changes

This is a re-land of https://github.com/NixOS/nixpkgs/pull/218044, but against `staging` instead of `master`.

- use cudaPackages instead of cudatoolkit (reduces download/closure size)
- set C/C++ compiler when building with CUDA to ensure NVCC has an appropriate backing compiler
- add flag to build with CUDNN (disabled by default due to increase in closure size)
- add flag to build with LTO (enabled by default)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
Running `nixpkgs-review` is taking a long time and seems to be stuck building one of the last six derivations. Here's the log so far: https://gist.github.com/ConnorBaker/bb6ae8dc865672e27e3b233defca5ffa

And the failures:

###### Failed derivations

- [ ] Mathematica_13.2.1_BNDL_LINUX.sh.drv (derivation hash: 5gr2xs65ckxmcs9axyd56mqcfrfvvsb7)
  - https://gist.github.com/bd6150f05d221c520f2cf870f22853f1
- [ ] WolframEngine_13.1.0_LINUX.sh.drv (derivation hash: pxzjd4np0yg68yxfranzhzp2f193wcv1)
  - https://gist.github.com/54a3efd6d6ad5b8ae19933dc66d286e5
- [ ] ibtws_9542.jar.drv (derivation hash: 6sfi97mppf0qfiqy79xaxqhby298dva5)
  - https://gist.github.com/6106aa4f6935479e144889b71c4ba173
- [ ] jdk-8u281-linux-x64.tar.gz.drv (derivation hash: vijjp643q981c81g57imy2d0cdw2pi1z)
  - https://gist.github.com/71c91500f1caa5ee44dd4933585bb857
- [ ] linuxx64-23.2.0.10.tar.gz.drv (derivation hash: xdpw5jy0ld7m3wc4f0sgcah2ccg9kq2k)
  - https://gist.github.com/8f36d2ce213b8260eef58ae3b5b0af86
- [ ] libarchive-3.6.2.drv (derivation hash: 448dh2vrphvp6g9savj72kwjkj1zd2n4)
  - https://gist.github.com/6d0de3d77862014f36ed016909fa84fb
- [ ] python3.10-terminado-0.17.1.drv (derivation hash: jd3a8j5rwilr3gs50w9x7xd6mv2mvvvd)
  - https://gist.github.com/b3fbdaa9916af4b9d3b6d6d609030884
- [ ] orc-0.4.33.drv (derivation hash: q8srf5bssf2435swvi8h1jfc13a845vx)
  - https://gist.github.com/4de542f59d92067fe9893ea7ab611e5a
- [ ] libarchive-3.6.2.drv (derivation hash: 0zdn971xpp21isi1v9q7j2n7bp8igfmv)
  - https://gist.github.com/2d832106eee7c674e4e196d64e7e1098
- [ ] xmlsec-1.2.34.drv (derivation hash: sfq1syg7vlds2vqp0byzyw93faxcjxw4)
  - https://gist.github.com/2940c21bdb02d61ef2f2fe5f689a53bb
- [ ] openldap-2.6.4.drv (derivation hash: rwb7d44vcrlwmz901h045bs5zjdc3jql)
  - https://gist.github.com/1a04cd6d6d88ad4be2bb2117b64a335c

**Review by samuela — APPROVED** [review]

File: pkgs/development/libraries/opencv/4.x.nix [hunk 1]
```diff
@@ -227,6 +229,33 @@ let
...
+  ] ++ lib.optionals enableCudnn [
+    cudnn # cudnn.h
+  ] ++ lib.optionals enableCufft [
+    libcufft # cufft.h
+  ];
+
+  cuda-native-redist = symlinkJoin {
+    name = "cuda-native-redist-${cudaVersion}";
+    paths = with cudaPackages; [
+      cuda_cudart # cuda_runtime.h
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/opencv/4.x.nix:250]: Does anything break if we put `cuda_cudart` in `buildInputs` instead? As such, it's a shared library used at runtime

File: pkgs/development/libraries/opencv/4.x.nix [hunk 2]
```diff
@@ -338,12 +368,30 @@ stdenv.mkDerivation {
...
+        stdenv.cc.isClang ||
+          # or through cudatoolkit.
+          (enableCuda && cudatoolkit.cc.isClang)
+      )
+    ))
   ] ++ lib.optionals enableCuda [
     "-DCUDA_FAST_MATH=ON"
-    "-DCUDA_HOST_COMPILER=${cudatoolkit.cc}/bin/cc"
+    # We need to set the C and C++ host compilers for CUDA to the same compiler.
+    "-DCMAKE_C_COMPILER=${cudatoolkit.cc}/bin/cc"
```
[the remark this answers is not in the document]
Comment (ConnorBaker) [reply pkgs/development/libraries/opencv/4.x.nix:393]: Good catch! Will fix.

**Comment by ConnorBaker** [comment]
Hi all,

Are there any blockers preventing this from being merged?

**Comment by samuela** [comment]
LGTM, @basvandijk @SomeoneSerge thoughts?

File: pkgs/development/libraries/opencv/4.x.nix [hunk 3]
```diff
@@ -338,12 +368,30 @@ stdenv.mkDerivation {
...
+    (opencvFlag "CUFFT" enableCufft)
+
+    # LTO options
+    (opencvFlag "ENABLE_LTO" enableLto)
+    (opencvFlag "ENABLE_THIN_LTO" (
+      enableLto && (
+        # Only clang supports thin LTO, so we must either be using clang through the stdenv,
+        stdenv.cc.isClang ||
+          # or through the backend stdenv.
+          (enableCuda && backendStdenv.cc.isClang)
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/opencv/4.x.nix:387]: I guess we currently can't test this as we hard-wire cuda to gcc. H'm

**Review by SomeoneSerge — APPROVED** [review]

**Comment by samuela** [comment]
In the interest of keeping PRs moving along, I went ahead and merged! Thanks so much @ConnorBaker !

**Pull request merged**